File: src/main/java/co/nlighten/jsontransform/functions/common/CompareBy.java
package co.nlighten.jsontransform.functions.common;

import co.nlighten.jsontransform.adapters.JsonAdapter;

import java.util.Comparator;
import java.util.List;
import java.util.Objects;

public class CompareBy<JE> {
    public List<JE> by;
    public JE value;
    public CompareBy(JE value) {
        this.value = value;
    }

    public static <JE, JA extends Iterable<JE>, JO extends JE> Comparator<CompareBy<JE>> createByComparator(
            JsonAdapter<JE, JA, JO> adapter, int index, String type) {
        Comparator<CompareBy<JE>> comparator;
        if (type == null || "AUTO".equalsIgnoreCase(type)) {
            comparator = (CompareBy<JE> a, CompareBy<JE> b) -> {
                var compareResult = adapter.compareTo(a.by.get(index), b.by.get(index));
                return Objects.requireNonNullElse(compareResult, 0);
            };
        } else {
            comparator = switch (type.toUpperCase()) {
                case "NUMBER" -> Comparator.comparing((CompareBy<JE> tup) -> adapter.getNumberAsBigDecimal(tup.by.get(index)));
                case "BOOLEAN" -> Comparator.comparing((CompareBy<JE> tup) -> adapter.getBoolean(tup.by.get(index)));
                //case "STRING"
                default -> Comparator.comparing((CompareBy<JE> tup) -> {
                    var val = tup.by.get(index);
                    // null keys and empty string are treated the same
                    return adapter.isNull(val) ? "" : adapter.getAsString(val);
                });
            };
        }
        return comparator;
    }
}

File: javascript/json-transform/src/functions/common/CompareBy.ts
import { Comparator, ComparatorFactory } from "@wortise/sequency";
import { compareTo, getAsString, isNullOrUndefined, isNumberType } from "../../JsonHelpers";
import { BigDecimal } from "./FunctionHelpers";

const factory = new ComparatorFactory<CompareBy>();

class CompareBy {
  public by: any[];
  public value: any;

  constructor(value: any) {
    this.value = value;
    this.by = [];
  }

  public static createByComparator(index: number, type: string | null): Comparator<CompareBy> {
    let comparator: Comparator<CompareBy>;
    if (isNullOrUndefined(type) || "AUTO" === type.toUpperCase()) {
      comparator = factory.compare((a, b) => {
        const compareResult = compareTo(a.by?.[index], b.by?.[index]);
        return compareResult ?? 0;
      });
    } else {
      switch (type.toUpperCase()) {
        case "NUMBER": {
          comparator = factory.compare((_a, _b) => {
            const a = _a.by?.[index],
              b = _b.by?.[index];
            if ((isNumberType(a) || typeof a === "string") && (isNumberType(b) || typeof b === "string")) {
              return BigDecimal(a).comparedTo(BigDecimal(b));
            } else if (isNullOrUndefined(a) && !isNullOrUndefined(b)) {
              return -1;
            } else if (!isNullOrUndefined(a) && isNullOrUndefined(b)) {
              return 1;
            }
            return 0;
          });
          break;
        }
        case "BOOLEAN": {
          comparator = factory.compare((_a, _b) => {
            const a = _a.by?.[index],
              b = _b.by?.[index];
            if (typeof a === "boolean" && typeof b === "boolean") {
              return a === b ? 0 : a ? 1 : -1;
            } else if (isNullOrUndefined(a) && !isNullOrUndefined(b)) {
              return -1;
            } else if (!isNullOrUndefined(a) && isNullOrUndefined(b)) {
              return 1;
            }
            return 0;
          });
          break;
        }
        //case "STRING"
        default: {
          comparator = factory.compareBy(tup => getAsString(tup.by?.[index]) ?? "");
          break;
        }
      }
    }
    return comparator;
  }
}

export default CompareBy;
